### DISCUSSION TITLE
How to correctly close an async Connection created in a child Thread?

### DISCUSSION THREAD

**Discussion by vladislav-krastev** [sqlalchemy/sqlalchemy#12152] [Usage Questions] [answered]
Hello, everyone!

So, I've been having an awesome experience with **SQLAlchemy** so far, but I just can't seem to wrap my mind around some of the ways it works.

I'm creating an async `engine` with the `asyncpg` driver **_inside_** the `__init__()` of a class. To validate the _connection_string_ before allowing the app (a `FastAPI` api with `uvicorn`) to continue, I'm using the following logic, still **_inside_** the same `__init__()` method (that's way the ugly things with the `Thread`):
```python
import asyncio
import threading

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine

class DB:
    def __init__(self) -> None:
        engine = create_async_engine(
            "some connection string",
            pool_pre_ping=True,
        )

        error = None
        async def validate():
            try:
                async with engine.connect() as conn:
                    if len((await conn.execute(text("SELECT 1;"))).all()) == 1: ...
            except Exception as err:
                nonlocal error
                error = err
        thread = threading.Thread(target=asyncio.run, args=(validate(),))
        thread.start()
        thread.join()
        if error:
            print(f"Could not establish connection to {engine.url}: {error}")
            exit(1)
        print(f"established connection to {engine.url}")

        self._db = async_sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)
        print(engine.pool.status())
        return
```

The `print()` statement at the end keeps printing `Pool size: 5  Connections in pool: 1 Current Overflow: -4 Current Checked out connections: 0` no matter what I try, and later, when I actually try to use the `AssyncSession` in another method of the class, like let's say:
```python
    async def create(self):
        async with self._db() as db, db.begin():
            db.add(...)
        return
```

I get one `RuntimeWarning` and one `RuntimeError` at once:

The `RuntimeWarning`:
```
.../asyncpg/connection.py:1673: RuntimeWarning: coroutine 'Connection._cancel' was never awaited
  self._cancellations.add(self._loop.create_task(self._cancel(waiter)))
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Exception terminating connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7faa0def6990>>
```

The `RuntimeError`:
```
.../asyncpg/connection.py", line 1673, in _cancel_current_command
    self._cancellations.add(self._loop.create_task(self._cancel(waiter)))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1427, in uvloop.loop.Loop.create_task
  File "uvloop/loop.pyx", line 705, in uvloop.loop.Loop._check_closed
RuntimeError: Event loop is closed
```

Tried any kind of combinations of:
```python
await conn.commit()
await conn.invalidate()
await conn.close()
engine.pool.dispose()
engine.pool = engine.pool.recreate()
```
inside the `try:` block of the  `async def validate()` function, but the `print()` statement at the end of the `__init__()` always keeps printing the same message and the two errors keep showing up.

**_ANY_** kind of suggestions or pointing me to what I'm doing wrong or what should I refer from the docs?

Big _thank you in advance_ to everyone and have a great day!

**Comment by CaselIT** [comment, marked as answer]
Hi,

don't share async things between threads. Each threads has its own event loop and in general you can't mix different event loops.

Since you mention an asgi application, use the lifespan events to initialize the db, they are there for that reason. the startup one seems to be the one you want https://asgi.readthedocs.io/en/latest/specs/lifespan.html I don't know how fastapi exposes them, but I think they should be available

**Reply by vladislav-krastev** [reply]
Hello, and thank you for the quick answer!

To be clear - you _**can**_ or you _**can't**_  "mix" event loops in general?
My issue is not that I'm doing something specifically `SQLAlchemy`-wrong, it's that the "main thread -> event loop in main thread -> child thread -> event loop in child thread -> event loop in main thread" flow is just incorrect/not-supported/bad idea?

Am I understanding you correctly?

**Reply by CaselIT** [reply]
> To be clear - you can or you can't "mix" event loops in general?

can't, sorry. fixed the message

> Am I understanding you correctly?

yes.

my suggestion would be to use one of the event mentioned above or use a different engine in the thread that you dispose before exiting the thread

**Reply by vladislav-krastev** [reply]
Awesome!
Thank you again!!!
